File: crates/h5i-db-capture/src/writer.rs
//! Hourly lz4 NDJSON files, rolled by arrival time.
//!
//! Layout, chosen so a partial capture is still navigable and an hour is a
//! unit you can copy, delete or re-ingest on its own:
//!
//! ```text
//! <root>/<venue>/<YYYY-MM-DD>/<HH>.ndjson.lz4
//! <root>/<venue>/<YYYY-MM-DD>/<HH>.001.ndjson.lz4   (second run in that hour)
//! ```
//!
//! # One lz4 frame per file, and why there is a part number
//!
//! `lz4_flex`'s decoder stops at the first frame's end mark: a file holding
//! two concatenated frames reads back as only its first frame, silently. So a
//! recorder restarted inside an hour it has already written cannot append,
//! and must not truncate either. It opens the next part instead, and a reader
//! globs `<HH>*.ndjson.lz4` to get the hour whole.

use std::fs::{File, OpenOptions};
use std::io::{self, BufWriter, Write};
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

use lz4_flex::frame::FrameEncoder;

/// Writes archive lines to the file for the hour each line arrived in.
pub struct CaptureWriter {
    root: PathBuf,
    venue: String,
    flush_after: Duration,
    open: Option<OpenHour>,
    lines: u64,
}

struct OpenHour {
    hour: i64,
    path: PathBuf,
    /// `Option` only so `close` can move the encoder out to finish it.
    encoder: Option<FrameEncoder<BufWriter<File>>>,
    last_flush: Instant,
    unflushed: u64,
}

impl CaptureWriter {
    /// `flush_after` bounds how much an ungraceful death costs.
    ///
    /// Bytes sit in the lz4 block buffer and then in the `BufWriter` until
    /// something pushes them out, and `kill -9` gets no chance to. Flushing on
    /// an interval puts every completed block on disk, so a killed recorder
    /// leaves a frame that is missing only its end mark: a tolerant reader
    /// still recovers every complete block, which is nearly everything.
    pub fn new(root: impl Into<PathBuf>, venue: impl Into<String>, flush_after: Duration) -> Self {
        Self {
            root: root.into(),
            venue: venue.into(),
            flush_after,
            open: None,
            lines: 0,
        }
    }

    /// Lines written since the writer was created.
    pub fn lines(&self) -> u64 {
        self.lines
    }

    /// The file currently being written, if any.
    pub fn current_path(&self) -> Option<&Path> {
        self.open.as_ref().map(|open| open.path.as_path())
    }

    /// Write one line, rolling the file if `received_at` is in a new hour.
    ///
    /// Rolling on *arrival* time rather than venue time keeps the file name a
    /// true statement about the file: everything in `14.ndjson.lz4` was seen
    /// between 14:00 and 15:00 UTC, whatever the payloads claim.
    pub fn write_line(&mut self, received_at: u64, line: &str) -> io::Result<()> {
        let hour = (received_at / 1_000_000_000) as i64 / 3600;
        if self.open.as_ref().is_some_and(|open| open.hour != hour) {
            self.close()?;
        }
        if self.open.is_none() {
            self.open = Some(self.open_hour(hour)?);
        }
        let flush_after = self.flush_after;
        let open = self
            .open
            .as_mut()
            .expect("an hour was just opened if it was missing");
        let encoder = open
            .encoder
            .as_mut()
            .expect("the encoder is only absent inside close");
        encoder.write_all(line.as_bytes())?;
        encoder.write_all(b"\n")?;
        open.unflushed += 1;
        self.lines += 1;
        if open.last_flush.elapsed() >= flush_after {
            self.flush()?;
        }
        Ok(())
    }

    /// Push completed blocks down to the file.
    ///
    /// Called on the flush interval and from the shutdown path. A flush with
    /// nothing pending is skipped, so a quiet market costs no syscalls.
    pub fn flush(&mut self) -> io::Result<()> {
        let Some(open) = self.open.as_mut() else {
            return Ok(());
        };
        if open.unflushed == 0 {
            open.last_flush = Instant::now();
            return Ok(());
        }
        let encoder = open
            .encoder
            .as_mut()
            .expect("the encoder is only absent inside close");
        // Two flushes: the first ends the lz4 block, the second moves the
        // resulting bytes out of the BufWriter and into the file.
        encoder.flush()?;
        encoder.get_mut().flush()?;
        open.last_flush = Instant::now();
        open.unflushed = 0;
        Ok(())
    }

    /// Write the frame's end mark, then flush and fsync the file.
    ///
    /// The end mark is what makes the file decodable at all, and the fsync is
    /// why shutdown is worth handling rather than leaving to the OS: without
    /// it a machine that loses power just after SIGINT still loses the tail
    /// the operator watched being written.
    pub fn close(&mut self) -> io::Result<()> {
        let Some(mut open) = self.open.take() else {
            return Ok(());
        };
        let encoder = open
            .encoder
            .take()
            .expect("the encoder is only absent inside close");
        let mut sink = encoder.finish().map_err(io::Error::other)?;
        sink.flush()?;
        sink.into_inner()
            .map_err(|error| io::Error::other(error.to_string()))?
            .sync_all()
    }

    fn open_hour(&self, hour: i64) -> io::Result<OpenHour> {
        let stamp = chrono::DateTime::from_timestamp(hour * 3600, 0)
            .ok_or_else(|| io::Error::other(format!("hour {hour} is not a representable time")))?;
        let dir = self
            .root
            .join(&self.venue)
            .join(stamp.format("%Y-%m-%d").to_string());
        std::fs::create_dir_all(&dir)?;
        let hour_name = stamp.format("%H").to_string();
        // `create_new` rather than a stat-then-open: two recorders pointed at
        // one directory should collide loudly here rather than interleave two
        // frames into one unreadable file.
        for part in 0..=999u32 {
            let path = dir.join(match part {
                0 => format!("{hour_name}.ndjson.lz4"),
                // Zero padded so lexical order stays chronological order.
                _ => format!("{hour_name}.{part:03}.ndjson.lz4"),
            });
            match OpenOptions::new().create_new(true).write(true).open(&path) {
                Ok(file) => {
                    return Ok(OpenHour {
                        hour,
                        path,
                        encoder: Some(FrameEncoder::new(BufWriter::new(file))),
                        last_flush: Instant::now(),
                        unflushed: 0,
                    });
                }
                Err(error) if error.kind() == io::ErrorKind::AlreadyExists => continue,
                Err(error) => return Err(error),
            }
        }
        Err(io::Error::other(format!(
            "{} already has 1000 parts for hour {hour_name}",
            dir.display()
        )))
    }
}

impl Drop for CaptureWriter {
    fn drop(&mut self) {
        // A panic elsewhere should still leave a decodable file. Errors here
        // have nowhere to go, so the explicit `close` on the shutdown path is
        // what reports them; this is only the backstop.
        let _ = self.close();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Read;

    fn decompress(path: &Path) -> String {
        let bytes = std::fs::read(path).unwrap();
        let mut out = String::new();
        lz4_flex::frame::FrameDecoder::new(&bytes[..])
            .read_to_string(&mut out)
            .unwrap();
        out
    }

    #[test]
    fn lines_land_in_the_hour_they_arrived_in() {
        let dir = tempfile::tempdir().unwrap();
        let mut writer = CaptureWriter::new(dir.path(), "kalshi", Duration::from_secs(3600));
        // 2025-01-01T00:00:02Z and 2025-01-01T01:00:02Z.
        writer.write_line(1_735_689_602_000_000_000, "a").unwrap();
        writer.write_line(1_735_693_202_000_000_000, "b").unwrap();
        writer.close().unwrap();

        let day = dir.path().join("kalshi").join("2025-01-01");
        assert_eq!(decompress(&day.join("00.ndjson.lz4")), "a\n");
        assert_eq!(decompress(&day.join("01.ndjson.lz4")), "b\n");
        assert_eq!(writer.lines(), 2);
    }

    #[test]
    fn flushing_mid_file_leaves_one_frame_that_still_reads_whole() {
        // A zero flush interval flushes after every line. If flushing ever
        // starts a second frame instead of a second block, the decoder stops
        // at the first line and this test says so.
        let dir = tempfile::tempdir().unwrap();
        let mut writer = CaptureWriter::new(dir.path(), "polymarket", Duration::from_secs(0));
        for i in 0..4 {
            writer
                .write_line(1_735_689_602_000_000_000, &format!("line{i}"))
                .unwrap();
        }
        writer.close().unwrap();
        let path = dir.path().join("polymarket/2025-01-01/00.ndjson.lz4");
        assert_eq!(decompress(&path), "line0\nline1\nline2\nline3\n");
    }

    #[test]
    fn a_restart_inside_the_same_hour_opens_the_next_part() {
        // Appending would produce a two-frame file whose second frame the
        // decoder skips without complaint, which is the worst kind of loss.
        let dir = tempfile::tempdir().unwrap();
        for line in ["first", "second"] {
            let mut writer = CaptureWriter::new(dir.path(), "kalshi", Duration::from_secs(3600));
            writer.write_line(1_735_689_602_000_000_000, line).unwrap();
            writer.close().unwrap();
        }
        let day = dir.path().join("kalshi/2025-01-01");
        assert_eq!(decompress(&day.join("00.ndjson.lz4")), "first\n");
        assert_eq!(decompress(&day.join("00.001.ndjson.lz4")), "second\n");
    }

    #[test]
    fn a_killed_capture_still_yields_its_flushed_blocks() {
        // Simulates kill -9: flush, then never close, so the frame has no end
        // mark. The claim being tested is that the loss is bounded by the
        // flush interval rather than being the whole file.
        let dir = tempfile::tempdir().unwrap();
        let path = {
            let mut writer = CaptureWriter::new(dir.path(), "kalshi", Duration::from_secs(0));
            writer
                .write_line(1_735_689_602_000_000_000, "kept")
                .unwrap();
            let path = writer.current_path().unwrap().to_path_buf();
            // Leak rather than drop: `Drop` would close the frame properly.
            std::mem::forget(writer);
            path
        };
        let bytes = std::fs::read(&path).unwrap();
        let mut out = Vec::new();
        let mut decoder = lz4_flex::frame::FrameDecoder::new(&bytes[..]);
        // The read errors at the missing end mark, after handing back the
        // block that was flushed.
        let _ = decoder.read_to_end(&mut out);
        assert_eq!(String::from_utf8(out).unwrap(), "kept\n");
    }
}

File: crates/h5i-db-python/python/h5i_db/capture/_writer.py
"""Hourly lz4 NDJSON files, rolled by arrival time.

Layout, chosen so a partial capture is still navigable and an hour is a unit
you can copy, delete or re-ingest on its own:

```text
<root>/<venue>/<YYYY-MM-DD>/<HH>.ndjson.lz4
<root>/<venue>/<YYYY-MM-DD>/<HH>.001.ndjson.lz4   (second run in that hour)
```

# One lz4 frame per file, and why there is a part number

`lz4.frame.decompress` and `LZ4FrameDecompressor` both stop at the first
frame's end mark: a file holding two concatenated frames reads back as only its
first frame, with no error at all. So a recorder restarted inside an hour it
has already written cannot append, and must not truncate either. It opens the
next part instead, and a reader globs `<HH>*.ndjson.lz4` to get the hour whole.

This is also why the periodic flush uses `compress_flush(end_frame=False)` and
never `LZ4FrameCompressor.flush()`, which ends the frame. A flush that ended
the frame would make every flush interval the start of a new, invisible file.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, BinaryIO, Optional

from ._deps import require

__all__ = ["CaptureWriter", "read_capture", "read_capture_text", "read_hour"]

_NANOS_PER_SECOND = 1_000_000_000
_SECONDS_PER_HOUR = 3600
_EPOCH = datetime(1970, 1, 1, tzinfo=timezone.utc)

#: How many parts one hour may have before the recorder gives up. A restart
#: loop is the usual cause, and failing loudly beats filling a directory.
_MAX_PARTS = 1000


def _lz4() -> Any:
    return require("lz4.frame", why="write and read lz4-compressed captures")


@dataclass
class _OpenHour:
    """The file currently being written, plus the state to finish it."""

    hour: int
    path: Path
    handle: BinaryIO
    context: Any
    last_flush: float
    unflushed: int = 0


class CaptureWriter:
    """Writes archive lines to the file for the hour each line arrived in.

    `flush_after` bounds how much an ungraceful death costs. Bytes sit in the
    lz4 block buffer and then in the file object's buffer until something
    pushes them out, and `kill -9` gets no chance to. Flushing on an interval
    puts every completed block on disk, so a killed recorder leaves a frame
    that is missing only its end mark: :func:`read_capture` still recovers
    every complete block, which is nearly everything.
    """

    def __init__(
        self,
        root: str | os.PathLike[str],
        venue: str,
        flush_after: float = 5.0,
    ) -> None:
        self._root = Path(root)
        self._venue = venue
        self._flush_after = float(flush_after)
        self._open: Optional[_OpenHour] = None
        self._lines = 0

    @property
    def lines(self) -> int:
        """Lines written since the writer was created."""
        return self._lines

    @property
    def current_path(self) -> Optional[Path]:
        """The file currently being written, if any."""
        return self._open.path if self._open is not None else None

    def write_line(self, received_at: int, line: str) -> None:
        """Write one line, rolling the file if `received_at` is in a new hour.

        Rolling on *arrival* time rather than venue time keeps the file name a
        true statement about the file: everything in `14.ndjson.lz4` was seen
        between 14:00 and 15:00 UTC, whatever the payloads claim.
        """
        hour = int(received_at) // _NANOS_PER_SECOND // _SECONDS_PER_HOUR
        if self._open is not None and self._open.hour != hour:
            self.close()
        if self._open is None:
            self._open = self._open_hour(hour)
        open_hour = self._open
        lz4 = _lz4()
        payload = line.encode("utf-8") + b"\n"
        open_hour.handle.write(lz4.compress_chunk(open_hour.context, payload))
        open_hour.unflushed += 1
        self._lines += 1
        if time.monotonic() - open_hour.last_flush >= self._flush_after:
            self.flush()

    def flush(self) -> None:
        """Push completed blocks down to the file.

        Called on the flush interval and from the shutdown path. A flush with
        nothing pending is skipped, so a quiet market costs no syscalls.
        """
        open_hour = self._open
        if open_hour is None:
            return
        if open_hour.unflushed == 0:
            open_hour.last_flush = time.monotonic()
            return
        lz4 = _lz4()
        # `end_frame=False` is the whole point: it ends the current lz4 block
        # without writing the frame's end mark, so the file stays one frame.
        open_hour.handle.write(lz4.compress_flush(open_hour.context, end_frame=False))
        open_hour.handle.flush()
        open_hour.last_flush = time.monotonic()
        open_hour.unflushed = 0

    def close(self) -> None:
        """Write the frame's end mark, then flush and fsync the file.

        The end mark is what makes the file decodable by a strict reader at
        all, and the fsync is why shutdown is worth handling rather than
        leaving to the OS: without it a machine that loses power just after
        SIGINT still loses the tail the operator watched being written.

        Idempotent, so the shutdown path and the fallback in `__del__` can both
        call it.
        """
        open_hour = self._open
        if open_hour is None:
            return
        self._open = None
        lz4 = _lz4()
        try:
            open_hour.handle.write(lz4.compress_flush(open_hour.context, end_frame=True))
            open_hour.handle.flush()
            os.fsync(open_hour.handle.fileno())
        finally:
            open_hour.handle.close()

    def __enter__(self) -> "CaptureWriter":
        return self

    def __exit__(self, *_exc: object) -> None:
        self.close()

    def __del__(self) -> None:
        # A crash elsewhere should still leave a decodable file. Errors here
        # have nowhere to go, so the explicit `close` on the shutdown path is
        # what reports them; this is only the backstop.
        try:
            self.close()
        except Exception:  # pragma: no cover - interpreter teardown
            pass

    def _open_hour(self, hour: int) -> _OpenHour:
        stamp = _EPOCH + timedelta(seconds=hour * _SECONDS_PER_HOUR)
        directory = self._root / self._venue / stamp.strftime("%Y-%m-%d")
        directory.mkdir(parents=True, exist_ok=True)
        hour_name = stamp.strftime("%H")
        lz4 = _lz4()
        for part in range(_MAX_PARTS):
            name = (
                f"{hour_name}.ndjson.lz4"
                if part == 0
                # Zero padded so lexical order stays chronological order.
                else f"{hour_name}.{part:03d}.ndjson.lz4"
            )
            path = directory / name
            try:
                # O_EXCL rather than a stat-then-open: two recorders pointed at
                # one directory should collide loudly here rather than
                # interleave two frames into one silently truncated file.
                descriptor = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644)
            except FileExistsError:
                continue
            handle = os.fdopen(descriptor, "wb")
            context = lz4.create_compression_context()
            # Independent blocks, no checksums: the same frame options the Rust
            # writer used, and independence means a block damaged by a partial
            # write does not make the blocks after it unreadable too.
            handle.write(
                lz4.compress_begin(
                    context,
                    block_linked=False,
                    block_checksum=False,
                    content_checksum=False,
                    auto_flush=0,
                )
            )
            return _OpenHour(
                hour=hour,
                path=path,
                handle=handle,
                context=context,
                last_flush=time.monotonic(),
            )
        raise OSError(
            f"{directory} already has {_MAX_PARTS} parts for hour {hour_name}"
        )


def read_capture_text(
    path: str | os.PathLike[str], *, tolerant: bool = True
) -> str:
    """Decompress one capture file to its NDJSON text.

    `tolerant` is on by default because the interesting file is the one a
    `kill -9` left behind: it has every flushed block and no end mark, and a
    strict decode throws away the blocks along with the complaint. A truncated
    trailing line is dropped rather than returned half-read, since a reader
    that sees it will only fail to parse it.
    """
    lz4 = _lz4()
    data = Path(path).read_bytes()
    decompressor = lz4.LZ4FrameDecompressor()
    out = bytearray()
    # Fed in chunks so a failure part way through keeps what came before it.
    for start in range(0, len(data), 1 << 16):
        try:
            out += decompressor.decompress(data[start : start + (1 << 16)])
        except RuntimeError:
            if not tolerant:
                raise
            break
    if not decompressor.eof and not tolerant:
        raise ValueError(
            f"{path} has no lz4 end mark: it was written by a recorder that "
            "was killed. Read it with tolerant=True to recover the flushed "
            "blocks."
        )
    return bytes(out).decode("utf-8", "ignore" if tolerant else "strict")


def read_capture(
    path: str | os.PathLike[str], *, tolerant: bool = True
) -> list[dict[str, Any]]:
    """One capture file as parsed envelopes, in the order they were written."""
    text = read_capture_text(path, tolerant=tolerant)
    lines = text.split("\n")
    if lines and lines[-1] != "":
        # No trailing newline means the last line is half a line.
        lines.pop()
    return [json.loads(line) for line in lines if line]


def _part_number(path: Path) -> int:
    """The part index of `HH.ndjson.lz4` (0) or `HH.NNN.ndjson.lz4` (NNN)."""
    pieces = path.name.split(".")
    if len(pieces) >= 4 and pieces[1].isdigit():
        return int(pieces[1])
    return 0


def read_hour(
    directory: str | os.PathLike[str], hour: str, *, tolerant: bool = True
) -> list[dict[str, Any]]:
    """Every part of one hour, concatenated in part order.

    The part files exist because a restart cannot append, so reading an hour
    means reading all of them; a caller who globs `<HH>.ndjson.lz4` alone
    silently drops everything the restarted recorder wrote.

    Sorted by part *number*, not by name. Part zero has no suffix at all, so
    `sorted()` puts `00.001.ndjson.lz4` before `00.ndjson.lz4` and hands back an
    hour whose first file is its second run.
    """
    parts = sorted(Path(directory).glob(f"{hour}*.ndjson.lz4"), key=_part_number)
    out: list[dict[str, Any]] = []
    for part in parts:
        out.extend(read_capture(part, tolerant=tolerant))
    return out
